### PR TITLE
Refactor Template and improve performance and type safety

### PR DESCRIPTION
**Pull request by mvorisek**
marking as `BC break` as the internal structure is no longer accessible from outside the class

otherwise fully BC, this feature is not used much in a custom code

even with additional checks the whole app (unit tests) are overall 11% faster than before

before:
image

after:
image

### CODE REVIEW

**Comment by mvorisek** [comment]
@ibelar can you review this PR

File: src/Template.php [hunk 1]
```diff
@@ -120,24 +102,19 @@ public function isTopTag($tag): bool
...
-        $tag = array_shift($a);
-        //$ref = array_shift($a); // unused
+        $tag = explode('#', $tag, 2)[0];
         if (!isset($this->tags[$tag])) {
-            throw $this->exception('Tag not found in Template')
-                ->addMoreInfo('tag', $tag)
-                ->addMoreInfo('tags', implode(', ', array_keys($this->tags)));
+            throw $this->exceptionAddMoreInfo(
+                (new Exception('Tag not found in Template'))
+                    ->addMoreInfo('tag', $tag)
```
Comment (romaninsh) [inline src/Template.php]: listing existing tags is very helpful in the error message. Could we keep that?

File: src/Template.php [hunk 2]
```diff
@@ -154,51 +131,35 @@ public function &getTagRef($tag)
...
-        $ref = array_shift($a);
-        if (!$ref) {
+        if (strpos($tag, '#') === false) {
             if (!isset($this->tags[$tag])) {
-                throw $this->exception('Tag not found in Template')
-                    ->addMoreInfo('tag', $tag)
-                    ->addMoreInfo('tags', implode(', ', array_keys($this->tags)));
+                throw $this->exceptionAddMoreInfo(
+                    (new Exception('Tag not found in Template'))
+                        ->addMoreInfo('tag', $tag)
```
Comment (romaninsh) [inline src/Template.php]: same as above - tags.

File: src/Template.php [hunk 3]
```diff
@@ -662,23 +610,18 @@ public function tryLoad(string $filename)
...
-        $this->template = $this->tags = [];
-        if (!$str) {
-            return;
-        }
+        $this->tags = [];
+        $this->template = [];
         $this->tag_cnt = [];
 
-        // First expand self-closing tags {$tag} -> {tag}{/tag}
-        $str = preg_replace('/{\$([-_:\w]+)}/', '{\1}{/\1}', $str);
```
Comment (romaninsh) [inline src/Template.php:679]: moved to parseTemplate.

File: src/Template.php [hunk 4]
```diff
@@ -689,112 +632,86 @@ public function loadTemplateFromString($str)
...
-     *
-     * @return string unique tag name
+     * Register tag and return unique tag name.
      */
-    protected function regTag($tag)
+    protected function regTag(string $tag): string
     {
         if (!isset($this->tag_cnt[$tag])) {
-            $this->tag_cnt[$tag] = 0;
+            $this->tag_cnt[$tag] = -1;
```
Comment (romaninsh) [inline src/Template.php]: could we keep those at 0 and start tags with #1? may introduce a minor compatibility issue. I have used positional tag references in the past (although long time ago).

File: src/Template.php [hunk 5]
```diff
@@ -689,112 +632,86 @@ public function loadTemplateFromString($str)
...
      * Deploys parse recursion.
-     *
-     * @param string $str
      */
-    protected function parseTemplate($str)
+    protected function parseTemplate(string $str): void
     {
-        $tag = '/{([\/$]?[-_:\w]*[\?]?)}/';
+        $strHash = hash('sha256', $str);
+        if (!isset(self::$_parseTemplateCache[$strHash])) {
```
Comment (romaninsh) [inline src/Template.php]: clever ;)

**Review by romaninsh — CHANGES_REQUESTED** [review]
lets keep tag numbering as-is. Which part of this is BC break?

File: src/Template.php (hunk 1 above)
Comment (mvorisek) [reply src/Template.php]: `exceptionAddMoreInfo` does that... minor refactoring only.

File: src/Template.php (hunk 2 above)
Comment (mvorisek) [reply src/Template.php]: `exceptionAddMoreInfo` does that... minor refactoring only.

File: src/Template.php (hunk 3 above)
Comment (mvorisek) [reply src/Template.php:679]: yes...

File: src/Template.php (hunk 4 above)
Comment (mvorisek) [reply src/Template.php]: rebuildTags started on #0 already and we normalize it now to the more programmer way - the numbers after # are just like indexes! @romaninsh please rereview and approve

Comment (mvorisek) [reply src/Template.php]: closing this as this is now internal representaton only

File: src/Template.php [hunk 6]
```diff
@@ -18,191 +18,131 @@ class Template implements \ArrayAccess
...
             return [&$this->template];
         }
 
-        $a = explode('#', $tag);
-        $tag = array_shift($a);
-        $ref = array_shift($a);
-        if (!$ref) {
-            if (!isset($this->tags[$tag])) {
-                throw $this->exception('Tag not found in Template')
+        $tag = explode('#', $tag, 2)[0];
```
Comment (DarkSide666) [inline src/Template.php:132]: This code below slightly duplicate code in `getTagRef` method.

File: src/Template.php [hunk 7]
```diff
@@ -521,7 +466,7 @@ public function del($tag)
     public function tryDel($tag)
     {
         if (is_array($tag)) {
-            return $this->del($tag);
+            return $this->tryDel($tag);
```
Comment (DarkSide666) [inline src/Template.php]: This is wrong and will create infinite loop.
Probably should be:
```
if (is_array($tag)) {
    foreach ($tag as $t) {
        $this->tryDel($t);
    }

    return $this;
}
```

File: src/Template.php (hunk 6 above)
Comment (mvorisek) [reply src/Template.php:132]: Yes, but we need to access `tagsIndex` first. Later, we should probably store values of template structure in 1D array to fix the rest performance issues. (I expect like +200% faster code)

File: src/Template.php [hunk 8]
```diff
@@ -594,24 +527,22 @@ public function eachTag($tag, $callable)
...
-     * @return self
+     * @return static
      */
-    public function cloneRegion($tag)
+    public function cloneRegion(string $tag)
     {
-        if ($this->isTopTag($tag)) {
-            return clone $this;
+        $template = new static();
+        $template->app = $this->app;
```
Comment (DarkSide666) [inline src/Template.php:539]: Is App required for template to work? If it's not mandatory, then this line should be optional.

**Review by DarkSide666 — CHANGES_REQUESTED** [review]
Found 1 bug, but besides that LGTM.

File: src/Template.php (hunk 7 above)
Comment (mvorisek) [reply src/Template.php]: good catch, fixed

File: src/Template.php (hunk 8 above)
Comment (mvorisek) [reply src/Template.php:539]: see Skype, was needed and is needed, if we want to not change other core things (Model or how templates can be used)

**Comment by romaninsh** [comment]
@mvorisek are you planning to revert on the numbering from 0 for better BC?

File: src/Template.php [hunk 9]
```diff
@@ -689,112 +614,87 @@ public function loadTemplateFromString($str)
...
-     * @return string unique tag name
+     * Register tag and return unique tag name.
      */
-    protected function regTag($tag)
+    protected function regTag(string $tag): string
     {
-        if (!isset($this->tag_cnt[$tag])) {
-            $this->tag_cnt[$tag] = 0;
+        if (!isset($this->tagCnt[$tag])) {
+            $this->tagCnt[$tag] = -1;
```
Comment (romaninsh) [inline src/Template.php:630]: can we keep this as 0 for BC?

**Review by romaninsh — CHANGES_REQUESTED** [review]

File: src/Template.php (hunk 9 above)
Comment (mvorisek) [reply src/Template.php:630]: no longer avaiable to outside, thus irrelevant

**Review by DarkSide666 — APPROVED** [review]

**Pull request merged**